File: backend/app/api/favorites.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from typing import Optional

from app.core.database import get_db
from app.core.deps import get_current_active_user
from app.models.user import User
from app.models.favorite import Favorite
from app.models.generation import Generation
from app.schemas.generation import GenerationIdAction, GenerationOut, GenerationPage, NodeTypeCount
from app.api.generations import _get_any_generation, _to_out

router = APIRouter(prefix="/api/favorites", tags=["favorites"])


@router.post("", response_model=GenerationOut)
def add_favorite(
    payload: GenerationIdAction,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """收藏某条生成记录（可收藏画廊中他人的作品）。幂等：重复收藏直接返回。"""
    gen = _get_any_generation(payload.generation_id, db)
    existing = (
        db.query(Favorite)
        .filter(
            Favorite.user_id == current_user.id,
            Favorite.generation_id == gen.id,
        )
        .first()
    )
    if not existing:
        try:
            db.add(Favorite(user_id=current_user.id, generation_id=gen.id))
            db.commit()
        except IntegrityError:
            # 并发重复收藏时命中唯一约束，回滚后按已收藏处理
            db.rollback()
        db.refresh(gen)
    return _to_out(gen, current_user, include_username=True)


@router.get("", response_model=GenerationPage)
def list_favorites(
    keyword: Optional[str] = None,
    node_type: Optional[str] = None,
    skip: int = 0,
    limit: int = 20,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """当前用户的收藏列表（含生成记录内容，可按 keyword / node_type 筛选，分页返回）。"""
    query = db.query(Favorite).filter(Favorite.user_id == current_user.id)
    if keyword or node_type:
        query = query.join(Generation, Favorite.generation_id == Generation.id)
    if keyword:
        query = query.filter(Generation.name.ilike(f"%{keyword}%"))
    if node_type:
        query = query.filter(Generation.node_type == node_type)
    total = query.count()
    page_limit = min(limit, 100)
    favs = (
        query.order_by(Favorite.created_at.desc(), Favorite.id.desc())
        .offset(skip)
        .limit(page_limit)
        .all()
    )
    # 类型筛选项：收藏全部记录按类型分组计数（不受 keyword/node_type 过滤影响）
    node_type_counts = [
        NodeTypeCount(node_type=row[0], count=row[1])
        for row in db.query(Generation.node_type, func.count(Generation.id))
        .join(Favorite, Favorite.generation_id == Generation.id)
        .filter(Favorite.user_id == current_user.id)
        .group_by(Generation.node_type)
        .all()
    ]
    return GenerationPage(
        items=[_to_out(fav.generation, current_user, include_username=True) for fav in favs if fav.generation],
        total=total,
        skip=skip,
        limit=page_limit,
        node_type_counts=node_type_counts,
    )


@router.delete("/{generation_id}", response_model=GenerationOut)
def remove_favorite(
    generation_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """取消收藏。"""
    fav = (
        db.query(Favorite)
        .filter(
            Favorite.user_id == current_user.id,
            Favorite.generation_id == generation_id,
        )
        .first()
    )
    if fav:
        db.delete(fav)
        db.commit()
    gen = db.query(Generation).filter(Generation.id == generation_id).first()
    if not gen:
        raise HTTPException(status_code=404, detail="生成记录不存在")
    return _to_out(gen, current_user, include_username=True)

File: backend-ts/src/api/library/favorites.ts
import { and, count, desc, eq, like } from 'drizzle-orm';
import type { FastifyInstance } from 'fastify';
import { getDb } from '../../config/database.js';
import { favorites, generations } from '../../db/schema.js';
import { now } from '../../shared/datetime.js';
import { toGenerationOut, type GenerationPage, type GenerationRow } from './generations.js';

/**
 * 收藏接口（对应 Python `app/api/favorites.py`）：
 * - POST /api/favorites（收藏，可收藏画廊中他人作品，幂等）
 * - GET /api/favorites（当前用户收藏列表，keyword / node_type / 分页）
 * - DELETE /api/favorites/:generation_id（取消收藏）
 */

export async function registerFavoritesRouter(app: FastifyInstance): Promise<void> {
  // 收藏（幂等）
  app.post(
    '/api/favorites',
    { preHandler: app.authenticate },
    async (request, reply) => {
      const genId = Number((request.body as { generation_id?: number } | undefined)?.generation_id);
      const db = getDb();
      const gen = db.select().from(generations).where(eq(generations.id, genId)).get();
      if (!gen) return reply.code(404).send({ detail: '生成记录不存在' });
      const existing = db
        .select({ id: favorites.id })
        .from(favorites)
        .where(and(eq(favorites.userId, request.authUser!.id), eq(favorites.generationId, gen.id)))
        .get();
      if (!existing) {
        try {
          db.insert(favorites).values({ userId: request.authUser!.id, generationId: gen.id, createdAt: now() }).run();
        } catch {
          // 并发重复收藏命中唯一约束：按已收藏处理（无唯一索引时忽略即可）
        }
      }
      return toGenerationOut(db, gen, request.authUser!.id, true);
    }
  );

  // 收藏列表（含生成记录内容）
  app.get('/api/favorites', { preHandler: app.authenticate }, async (request) => {
    const q = (request.query ?? {}) as { keyword?: string; node_type?: string; skip?: string; limit?: string };
    const keyword = (q.keyword ?? '').trim();
    const nodeType = (q.node_type ?? '').trim();
    const skip = Number(q.skip ?? 0) || 0;
    const limit = Math.min(Number(q.limit ?? 20) || 20, 100);
    const db = getDb();
    const uid = request.authUser!.id;

    const filters = [eq(favorites.userId, uid)];
    if (keyword) filters.push(like(generations.name, `%${keyword}%`));
    if (nodeType) filters.push(eq(generations.nodeType, nodeType));
    const where = and(...filters);

    const total = db
      .select({ c: count() })
      .from(favorites)
      .innerJoin(generations, eq(favorites.generationId, generations.id))
      .where(where)
      .get()!.c;

    const rows = db
      .select({
        id: generations.id,
        userId: generations.userId,
        nodeType: generations.nodeType,
        name: generations.name,
        stageResults: generations.stageResults,
        resultUrl: generations.resultUrl,
        status: generations.status,
        createdAt: generations.createdAt,
      })
      .from(favorites)
      .innerJoin(generations, eq(favorites.generationId, generations.id))
      .where(where)
      .orderBy(desc(favorites.createdAt), desc(favorites.id))
      .limit(limit)
      .offset(skip)
      .all();

    const counts = db
      .select({ nodeType: generations.nodeType, c: count(generations.id) })
      .from(generations)
      .innerJoin(favorites, eq(favorites.generationId, generations.id))
      .where(eq(favorites.userId, uid))
      .groupBy(generations.nodeType)
      .all();

    return {
      items: rows.map((r) => toGenerationOut(db, r as GenerationRow, uid, true)),
      total,
      skip,
      limit,
      node_type_counts: counts.map((r) => ({ node_type: r.nodeType, count: r.c })),
    } satisfies GenerationPage;
  });

  // 取消收藏
  app.delete('/api/favorites/:generation_id', { preHandler: app.authenticate }, async (request, reply) => {
    const genId = Number((request.params as { generation_id: string }).generation_id);
    const db = getDb();
    db.delete(favorites)
      .where(and(eq(favorites.userId, request.authUser!.id), eq(favorites.generationId, genId)))
      .run();
    const gen = db.select().from(generations).where(eq(generations.id, genId)).get();
    if (!gen) return reply.code(404).send({ detail: '生成记录不存在' });
    return toGenerationOut(db, gen, request.authUser!.id, true);
  });
}
